onnxruntime/python/tools/quantization/calibrate.py

```diff
@@ -112,22 +211,62 @@ def get_intermediate_outputs(self, calib_mode='naive'):
         '''
 
         #conduct inference session and get intermediate outputs
-        session = onnxruntime.InferenceSession(self.augmented_model_path, None)
-
-        intermediate_outputs = []
-        while True:
-            inputs = self.data_reader.get_next()
-            if not inputs:
-                break
-            intermediate_outputs.append(session.run(None, inputs))
-        node_output_names = [session.get_outputs()[i].name for i in range(len(intermediate_outputs[0]))]
-        output_dicts_list = [
-            dict(zip(node_output_names, intermediate_output)) for intermediate_output in intermediate_outputs
-        ]
+        if providers:
+            sess_options = onnxruntime.SessionOptions()
+            sess_options.graph_optimization_level = onnxruntime.GraphOptimizationLevel.ORT_DISABLE_ALL #ORT_ENABLE_BASIC
+            session = onnxruntime.InferenceSession(self.augmented_model_path, sess_options=sess_options, providers=providers)
+        else:
+            session = onnxruntime.InferenceSession(self.augmented_model_path, None)
 
         #number of outputs in original model
         model = onnx.load(self.model_path)
         num_model_outputs = len(model.graph.output)
+        model_original_outputs = set(output.name for output in model.graph.output)
+
+        intermediate_outputs = []
+
+        if self.data_reader.support_batch_inference(session.get_inputs()[0].shape):
+            # batch inference
+            print("Doing batch inference.")
+            while True:
+                inputs = self.data_reader.get_batch()
+                if not inputs:
+                    break
+                intermediate_outputs.append(session.run(None, inputs))
+
+
+            node_output_names = [session.get_outputs()[i].name for i in range(len(intermediate_outputs[0]))]
+
+            output_dicts_list = []
+            output_dicts = {}
+            for intermediate_output in intermediate_outputs:
+                for i in range(len(intermediate_output)):
+                    if node_output_names[i] in model_original_outputs:
+                        continue
+
+                    if node_output_names[i] in output_dicts:
+                        if 'ReduceMin' in node_output_names[i]:
+                            output_dicts[node_output_names[i]] = min(output_dicts[node_output_names[i]], intermediate_output[i])
+                        elif 'ReduceMax' in node_output_names[i]:
+                            output_dicts[node_output_names[i]] = max(output_dicts[node_output_names[i]], intermediate_output[i])
+                    else:
+                        output_dicts[node_output_names[i]] = intermediate_output[i]
```

Comment:
you don't need this. it will be done below with code:

            pairs = [
                tuple([
                    float(min(clean_merged_dict[added_node_output_names[i]])),
                    float(max(clean_merged_dict[added_node_output_names[i + 1]]))
                ]) for i in range(0, len(added_node_output_names), 2)
            ]

Reply:
Yes, I will remove those lines.

Reply:
Also, the above code gets "only size-1 arrays can be converted to Python scalars" error when float(value) where value is [].

So, I made following change:
              pairs = []
              for i in range(0, len(added_node_output_names), 2):
                  min_value = 0
                  max_value = 0
                  min_value_array = min(clean_merged_dict[added_node_output_names[i]])
                  max_value_array = max(clean_merged_dict[added_node_output_names[i + 1]])
                  if type(min_value_array) == int or min_value_array.size > 0:
                      min_value = float(min_value_array)
                  if type(max_value_array) == int or max_value_array.size > 0:
                      max_value = float(max_value_array)

                  pairs.append(tuple([min_value, max_value]))